src/data/providers/initPoolDataProvider.ts

```diff
@@ -105,9 +105,39 @@ export class InitPoolDataProvider {
         poolAddress: Address,
         poolType: string,
     ): Promise<PoolState> {
-        console.log(poolAddress, poolType);
-        throw new Error(
-            'InitPoolData fetcher not implemented for Balancer V3 yet',
-        );
+        const chainId = await this.client.getChainId();
+        const vaultV3 = getContract({
+            abi: vaultV3Abi,
+            address: VAULT_V3[chainId],
+            client: this.client,
+        });
+
+        try {
+            const poolTokensFromVault: Address[] =
+                (await vaultV3.read.getPoolTokens([poolAddress])) as Address[];
+            const poolTokens = await Promise.all(
+                poolTokensFromVault.map(async (address, index) => {
+                    const decimals = await getTokenDecimals(
+                        address,
+                        this.client,
+                    );
+                    return {
+                        address: address.toLowerCase() as Address,
+                        index,
+                        decimals,
+                    };
+                }),
+            );
+            return {
+                id: poolAddress,
+                address: poolAddress.toLowerCase() as Address,
+                type: poolType,
+                tokens: poolTokens,
+                balancerVersion: 3,
+            };
+        } catch (e) {
+            console.warn(e);
+            throw new Error('Invalid pool address');
```

Comment:
I think this may be a bit misleading, since the error may also come from `getTokenDecimals`.
Ideally we should be able to remove the try/catch and rely on the errors coming from the contracts, but if they are not clear enough, maybe having one try/catch for each point of failure with a proper error message for each would make more sense.
What do you think?
Ps: this applies for the v2 implementation above
Ps2: getTokenDecimals could have an inner try/catch instead of having an external one here